// ng2-semantic-ui/components/select/multi-select-label.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, HostBinding, ElementRef, Renderer, HostListener, Input, ViewContainerRef, EventEmitter, ViewChild } from "@angular/core";
import { SuiTransition } from "../transition/transition";
export var SuiSelectMultiLabel = (function () {
    function SuiSelectMultiLabel(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.classes = true;
        this.useTemplate = false;
        this.readValue = function (value) { return ""; };
        this.selected = new EventEmitter();
        this._transition = new SuiTransition(el, renderer);
        this._transition.animate({
            name: "scale",
            direction: "in",
            display: "inline-block",
            duration: 100
        });
    }
    SuiSelectMultiLabel.prototype.click = function (event) {
        event.stopPropagation();
        return false;
    };
    SuiSelectMultiLabel.prototype.deselectOption = function () {
        var _this = this;
        event.stopPropagation();
        this._transition.animate({
            name: "scale",
            direction: "out",
            duration: 100,
            callback: function () { return _this.selected.emit(_this.value); }
        });
        return false;
    };
    __decorate([
        HostBinding('class.ui'),
        HostBinding('class.label'), 
        __metadata('design:type', Object)
    ], SuiSelectMultiLabel.prototype, "classes", void 0);
    __decorate([
        HostListener('click', ['$event']), 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [MouseEvent]), 
        __metadata('design:returntype', Boolean)
    ], SuiSelectMultiLabel.prototype, "click", null);
    __decorate([
        Input(), 
        __metadata('design:type', Object)
    ], SuiSelectMultiLabel.prototype, "value", void 0);
    __decorate([
        ViewChild('optionRenderTarget', { read: ViewContainerRef }), 
        __metadata('design:type', ViewContainerRef)
    ], SuiSelectMultiLabel.prototype, "viewContainerRef", void 0);
    SuiSelectMultiLabel = __decorate([
        Component({
            selector: 'sui-select-multi-label',
            template: "\n<span #optionRenderTarget></span>\n<span *ngIf=\"!useTemplate\">{{ readValue(value) }}</span>\n<i class=\"delete icon\" (click)=\"deselectOption()\"></i>\n"
        }), 
        __metadata('design:paramtypes', [ElementRef, Renderer])
    ], SuiSelectMultiLabel);
    return SuiSelectMultiLabel;
}());
